src/ClientTrait.php

```diff
@@ -0,0 +1,138 @@
+use Google\Auth\Credentials\GCECredentials;
+use Google\Auth\HttpHandler\HttpHandlerFactory;
+use Google\Cloud\Compute\Metadata;
+use Google\Cloud\Exception\ConfigException;
+use Google\Cloud\Exception\GoogleException;
+use GuzzleHttp\Psr7;
+
+trait ClientTrait
+{
+    /**
+     * @var string The project ID created in the Google Developers Console.
+     */
+    private $projectId;
+
+    /**
+     * Fetch and validate the keyfile and set the project ID.
+     *
+     * @param  array $config
+     * @return array
+     */
+    private function configureAuthentication(array $config)
+    {
+        $config['keyFile'] = $this->getKeyFile($config);
+
+        $this->projectId = $this->detectProjectId($config);
+
+        return $config;
+    }
+
+    /**
+     * Get a keyfile.
+     *
+     * Process:
+     * 1. If $config['keyFile'] is set, use that.
+     * 2. If $config['keyFilePath'] is set, load the file and use that.
+     * 3. If GOOGLE_APPLICATION_CREDENTIALS environment variable is set, load
+     *    from that location and use that.
+     * 4. If OS-specific well-known-file is set, load from that location and use
+     *    that.
+     * 5. Exception. :(
+     *
+     * @param  array $config
+     * @return array Key data
+     * @throws GoogleException
+     */
+    private function getKeyFile(array $config = [])
+    {
+        $config += [
+            'keyFile' => null,
+            'keyFilePath' => null,
+        ];
+
+        if ($config['keyFile']) {
+            return $config['keyFile'];
+        }
+
+        if ($config['keyFilePath']) {
+            if (!file_exists($config['keyFilePath'])) {
+                throw new GoogleException('Given keyfile path does not exist');
+            }
+
+            $keyFileData = json_decode(file_get_contents($config['keyFilePath']), true);
+
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                throw new GoogleException('Given keyfile was invalid');
+            }
+
+            return $keyFileData;
+        }
+
+        return CredentialsLoader::fromEnv()
+            ?: CredentialsLoader::fromWellKnownFile();
+    }
+
+    /**
+     * Detect and return a project ID.
+     *
+     * Process:
+     * 1. If $config['projectId'] is set, use that.
+     * 2. If $config['keyFile'] is set, attempt to retrieve a project ID from
+     *    that.
+     * 3. Throw exception.
+     *
+     * @param  array $config
+     * @return string
+     * @throws GoogleException
+     */
+    private function detectProjectId(array $config)
+    {
+        $config += [
+            'projectId' => null,
+            'httpHandler' => null,
+        ];
+
+        if ($config['projectId']) {
+            return $config['projectId'];
+        }
+
+        if (isset($config['keyFile']['project_id'])) {
+            return $config['keyFile']['project_id'];
+        }
+
+        $httpHandler = $config['httpHandler'] ? $config['httpHandler'] : null;
+        if (GCECredentials::onGce($httpHandler)) {
```

Comment:
What do you think about the following?

Nix line 124 and use:

`if (GCECredentials::onGce($config['httpHandler'])) {`

since the default is set above at the start of the method